### PR TITLE
Add support for file and folder operations (create, rename, move) to workspace edits

### PR DESCRIPTION
**Pull request by banacorn**
Instead of having `TextDocumentEdit` in `WorkspaceEdit`

```haskell
data WorkspaceEdit =
  WorkspaceEdit
    { _changes         :: Maybe WorkspaceEditMap
    , _documentChanges :: Maybe (List TextDocumentEdit)
    } deriving (Show, Read, Eq)
```

It is now replaced by a new type called `DocumentChange`

```haskell
data WorkspaceEdit =
  WorkspaceEdit
    { _changes         :: Maybe WorkspaceEditMap
    , _documentChanges :: Maybe (List DocumentChange)
    } deriving (Show, Read, Eq)
```

Which is just a synonym of union of `WorkspaceEdit` and other operations

```haskell
type DocumentChange = TextDocumentEdit |? CreateFile |? RenameFile |? DeleteFile
```

So that it complies with the [spec](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspaceEdit) as suggested by @EggBaconAndSpam in https://github.com/alanz/lsp/issues/134#issuecomment-502025882

---

Here's how to construct a `DocumentChange` that creates a file for example: 
```haskell
InR (InL (CreateFile FileResourceChangeCreate "/path/to/new/file" Nothing))
```

But it's a bit wordy and unsafe because of the `FileResoureChangeKind`.
I'm considering hiding these constructors and export only smart constructors.

### CODE REVIEW

**Comment by banacorn** [comment]
@bubba I should make a PR on [lsp-test](https://github.com/bubba/lsp-test), right? 
In that case, `lsp-test` would have to depend on this commit of `lsp` then 👀

**Comment by bubba** [comment]
Hi @banacorn, thanks for this PR! Looks like there's a bit of a chicken and the egg situation with lap-types and lsp-test. You can create a branch on lsp-test for now and reference that commit in `cabal.project`, and then once this is merged we can merge the PRs for lsp-test

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs [hunk 1]
```diff
@@ -38,12 +40,109 @@ deriveJSON lspOptions ''TextDocumentEdit
...
+  parseJSON (String "delete") = pure FileResourceChangeDelete
+  parseJSON _                 = mempty
+
+-- | Options to create a file.
+data CreateFileOptions =
+  CreateFileOptions
+    { -- | Overwrite existing file. Overwrite wins over `ignoreIfExists`
+      _overwrite      :: Bool
+      -- | Ignore if exists.
+    , _ignoreIfExists :: Bool
```
Comment (bubba) [inline lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:68]: To match the specification these should be maybes:

```typescript
/**
 * Options to create a file.
 */
export interface CreateFileOptions {
	/**
	 * Overwrite existing file. Overwrite wins over `ignoreIfExists`
	 */
	overwrite?: boolean;
	/**
	 * Ignore if exists.
	 */
	ignoreIfExists?: boolean;
}
```

```suggestion
    { -- | Overwrite existing file. Overwrite wins over `ignoreIfExists`
      _overwrite      :: Maybe Bool
      -- | Ignore if exists.
    , _ignoreIfExists :: Maybe Bool
```

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs [hunk 2]
```diff
@@ -38,12 +40,109 @@ deriveJSON lspOptions ''TextDocumentEdit
...
+
+deriveJSON lspOptions ''CreateFile
+
+-- Rename file options
+data RenameFileOptions =
+  RenameFileOptions
+    { -- | Overwrite target if existing. Overwrite wins over `ignoreIfExists`
+      _overwrite      :: Bool
+      -- | Ignores if target exists.
+    , _ignoreIfExists :: Bool
```
Comment (bubba) [inline lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:91]:
```suggestion
      _overwrite      :: Maybe Bool
      -- | Ignores if target exists.
    , _ignoreIfExists :: Maybe Bool
```

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs [hunk 3]
```diff
@@ -38,12 +40,109 @@ deriveJSON lspOptions ''TextDocumentEdit
...
+
+deriveJSON lspOptions ''RenameFile
+
+-- Delete file options
+data DeleteFileOptions =
+  DeleteFileOptions
+    { -- | Delete the content recursively if a folder is denoted.
+      _recursive          :: Bool
+      -- | Ignore the operation if the file doesn't exist.
+    , _ignoreIfNotExists  :: Bool
```
Comment (bubba) [inline lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:116]:
```suggestion
      _recursive          :: Maybe Bool
      -- | Ignore the operation if the file doesn't exist.
    , _ignoreIfNotExists  :: Maybe Bool
```

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs [hunk 4]
```diff
@@ -38,12 +40,109 @@ deriveJSON lspOptions ''TextDocumentEdit
...
+instance ToJSON FileResourceChangeKind where
+  toJSON FileResourceChangeCreate = String "create"
+  toJSON FileResourceChangeRename = String "rename"
+  toJSON FileResourceChangeDelete = String "delete"
+
+instance FromJSON FileResourceChangeKind where
+  parseJSON (String "create") = pure FileResourceChangeCreate
+  parseJSON (String "rename") = pure FileResourceChangeRename
+  parseJSON (String "delete") = pure FileResourceChangeDelete
+  parseJSON _                 = mempty
```
Comment (bubba) [inline lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:60]: Bit of a nitpick here, but I think we can avoid having to create this data type if we omit storing the `kind` in `{Create,Rename,Delete}File`. Instead we can just parse it and emit the `kind` field in custom `ToJSON` instances, like how `WorkDoneProgressBegin` is handled

https://github.com/alanz/lsp/blob/8d6e8f316a5a1c40986cdc6400611376c8ab0140/lsp-types/src/Language/LSP/Types/Progress.hs#L89-L107

I think this more closely matches the string literal tags in TypeScript interfaces, and it means users can't accidentally create a dodgy `CreateFile FileResourceChangeRename x`

**Review by bubba — COMMENTED** [review]
Thanks, this is great! Just some minor nitpicks about the string literal tag. Out of curiosity, are you using this for your own language server?

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs (hunk 1 above)
Comment (banacorn) [reply lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:68]: oopsy!

File: lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs (hunk 4 above)
Comment (banacorn) [reply lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs:60]: Nice!

**Comment by banacorn** [comment]
Yes, I'm building a language server for my toy language. 
I have a button that exports all proof obligations to a new file, that's why I need this feature.

I'm planning on building a language server for Agda next.

**Comment by bubba** [comment]
Whoops, looks like fix keyword in the PR message closes PRs across GitHub accounts

**Review by bubba — APPROVED** [review]
LGTM now. Let me know when you're happy/referencing the new lsp-test commit and I'll merge

**Comment by bubba** [comment]
Looks like its passing now. Want me to merge @banacorn ?

**Comment by banacorn** [comment]
Yes, please! Thank you!
